storage: add benchmark for DeepEqual(EvalResult)

tbg opened the pull request:
It showed up in profiles since it is in the hot path in `evaluateBatch` through
the call to `(*EvalResult).MergeAndDestroy` for each element of the batch.
That call to `reflect.DeepEqual` should be removed.

```
BenchmarkEquality/nil/builtin-4 	2000000000	         0.33 ns/op
BenchmarkEquality/nil/DeepEqual-4         	20000000	       101 ns/op
BenchmarkEquality/uint64x1/builtin-4      	2000000000	         0.32 ns/op
BenchmarkEquality/uint64x1/DeepEqual-4    	10000000	       190 ns/op
BenchmarkEquality/uint64x16/builtin-4     	100000000	        17.7 ns/op
BenchmarkEquality/uint64x16/DeepEqual-4   	 1000000	      1089 ns/op
BenchmarkEquality/uint1024x1/builtin-4    	100000000	        17.2 ns/op
BenchmarkEquality/uint1024x1/DeepEqual-4  	 1000000	      1117 ns/op
BenchmarkEquality/byteslicex1/DeepEqual-4 	 5000000	       247 ns/op
BenchmarkEquality/byteslicex16/DeepEqual-4         	 1000000	      1355 ns/op
BenchmarkEquality/EvalResult/DeepEqual-4           	  300000	      3968 ns/op
```

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/16569)
<!-- Reviewable:end -->

tbg committed:
storage: add benchmark for DeepEqual(EvalResult)

It showed up in profiles since it is in the hot path in `evaluateBatch` through
the call to `(*EvalResult).MergeAndDestroy` for each element of the batch.
That call to `reflect.DeepEqual` should be removed.

```
BenchmarkEquality/nil/builtin-4 	2000000000	         0.33 ns/op
BenchmarkEquality/nil/DeepEqual-4         	20000000	       101 ns/op
BenchmarkEquality/uint64x1/builtin-4      	2000000000	         0.32 ns/op
BenchmarkEquality/uint64x1/DeepEqual-4    	10000000	       190 ns/op
BenchmarkEquality/uint64x16/builtin-4     	100000000	        17.7 ns/op
BenchmarkEquality/uint64x16/DeepEqual-4   	 1000000	      1089 ns/op
BenchmarkEquality/uint1024x1/builtin-4    	100000000	        17.2 ns/op
BenchmarkEquality/uint1024x1/DeepEqual-4  	 1000000	      1117 ns/op
BenchmarkEquality/byteslicex1/DeepEqual-4 	 5000000	       247 ns/op
BenchmarkEquality/byteslicex16/DeepEqual-4         	 1000000	      1355 ns/op
BenchmarkEquality/EvalResult/DeepEqual-4           	  300000	      3968 ns/op
```

a-robinson commented on pkg/storage/bench_test.go:
```diff
@@ -0,0 +1,133 @@
+// implied. See the License for the specific language governing
+// permissions and limitations under the License.
+//
+// Author: Spencer Kimball (spencer.kimball@gmail.com)
```
I think these lines are pretty silly in general, but they're especially silly when you use someone else's name :)

a-robinson reviewed (approved):
LGTM, but do we gain anything from checking this in? It isn't benchmarking anything cockroach-specific that's likely to change over time.

tschottdorf commented:
Yeah, just wanted to toss this somewhere so I don't forget about it. Probably better kept on my repo though. I'll move this off when I get a chance next week. Thanks for taking a look!

tschottdorf commented:
Moved to github.com/tschottdorf/goplay/deep_equal_bench
